### PR TITLE
Added an extension point for custom function providers

### PR DESCRIPTION
**Pull request by zachmu**
As part of this, embarked on a major refactor:

* Extracted interfaces for sql.Catalog and sql.ProcessList
* Moved existing sql.Catalog to analyzer package
* Moved ProcessList and MemoryManager out of Catalog
* Changed Analyzer and Engine to take a DatabaseProvider instead of a Catalog

### CODE REVIEW

File: driver/conn.go [hunk 1]
```diff
@@ -23,17 +23,14 @@ import (
 
 // Conn is a connection to a database.
 type Conn struct {
-	options  Options
+	options  *Options
 	catalog  *catalog
```
Comment (andrew-wm-arthur) [inline driver/conn.go:27]: while you're here, `type catalog` on line 155 would benefit from having a new name

File: sql/expression/function/rpad_lpad.go [hunk 2]
```diff
@@ -33,11 +33,12 @@ const (
 	rPadType padType = 'r'
 )
 
-// NewPadFunc returns a Pad creator function with a specific padType.
-func NewPadFunc(pType padType) func(ctx *sql.Context, e ...sql.Expression) (sql.Expression, error) {
-	return func(ctx *sql.Context, e ...sql.Expression) (sql.Expression, error) {
-		return NewPad(ctx, pType, e...)
-	}
+func NewLeftPad(ctx *sql.Context, e ...sql.Expression) (sql.Expression, error) {
```
Comment (andrew-wm-arthur) [inline sql/expression/function/rpad_lpad.go:36]: don't break the internet Zach

**Review by andrew-wm-arthur — APPROVED** [review]
LGTM
Now that we've extracted an interface for `Catalog` I wonder about having `DatabaseProvider` as a separate concept. Feels like we could combine these

File: sql/expression/function/rpad_lpad.go (hunk 2 above)
Comment (zachmu) [reply sql/expression/function/rpad_lpad.go:36]: Don't ever tell me how to live my life

**Comment by zachmu** [comment]
> LGTM
> Now that we've extracted an interface for `Catalog` I wonder about having `DatabaseProvider` as a separate concept. Feels like we could combine these

Yeah, I think we're moving in that direction.

If anything, the Catalog concept is a caching layer for the analyzer, not a public interface. Just didn't want to go quite that far in one push.

**Pull request merged**